doc/guides/V8-api-for-node.md

```diff
@@ -0,0 +1,210 @@
+  XXX why are some called Error and others called Exception?
+  - SetMethod/SetMethodNoSideEffect/SetProtoMethod/SetTemplateMethod
+  XXX difference?
+  - SetImmediate
+  - EnvironmentOptions* options(): "some" options...
+  XXX how to reach PerIsolateOptions, PerProcessOptions
+  - etc.
+
+  Contains many global strings and symbols.
+  XXX ...
+
+
+## `NODE_MODULE_CONTEXT_AWARE_INTERNAL`
+
+See: https://nodejs.org/api/addons.html#addons_context_aware_addons
+
+Called with:
+- `Local<Object> exports`: where to put exported properties, conventionally
+  called `target` in node
+- `Local<Value> module`: conventionally unused in node
+XXX what is this for? addon docs don't mention or use it
+- `Local<Context> context`:
+- void* priv: not commonly used
+XXX where is it ever used? for what?
+
+Initialize is generally used to set methods and contants:
+
+        Environment* env = Environment::GetCurrent(context);
+        env->SetMethod(target, "name", Name);
+        // ... see Environment for method creation convenience functions
+        NODE_DEFINE_CONSTANT(target, MACRO_NAME);
+        // read-only, not deletable, not enumerable:
+        NODE_DEFINE_HIDDEN_CONSTANT(target, MACRO_NAME);
+
+For functions that wrap a C++ object in a js object, manually do what SetMethod
+does, see Hmac::Initialize as an example:
+1. Create a FunctionTemplate, used to setup function properties. The various
+wrappers all call v8::FunctionTemplate::New() with various arguments, but
+env->NewFunctionTemplate() is most commonly used. Signature,
+ConstructorBehavior, and SideEffectType can be customized, but aren't
+documented, and usually are left as default.
+2. Call env->setProtoMethod() to setup instance methods
+3. Get a function from the template
+4. Set a string in the target to the function
+
+Mysterious boilerplate:
+- ToLocalChecked() see MaybeLocal vs Local
+- FromJust(): XXX
+
+    
+    
+        Local<FunctionTemplate> t = env->NewFunctionTemplate(New);
+
+Fields are used to store pointers to C++ objects:
+XXX I think
+        t->InstanceTemplate()->SetInternalFieldCount(1);
+Set methods:
+
+
+https://v8.dev/docs/embed#more-example-code
+- XXX read through, it has examples of calling the API
+
+
+
+- <http://izs.me/v8-docs/classv8_1_1Object.html>
+- https://code.google.com/p/v8/
+- Building: <https://code.google.com/p/v8/wiki/BuildingWithGYP>
+- how to compile js to see what it looks like?
+- [Breaking V8 Changes](https://docs.google.com/a/strongloop.com/document/d/1g8JFi8T_oAE_7uAri7Njtig7fKaPDfotU6huOa1alds/edit)
+- <https://developers.google.com/v8/get_started>
+- <https://chromium.googlesource.com/v8/v8/+/master/docs/using_git.md>
+
+- <https://developers.google.com/v8/embed>
+
+Handle is base, from that are Local (go in HandleScope), and Persistent
+(manually managed scope). Constructors (String::New) seem to return Locals.
+
+`return Local<Array>();` ... seems to do exactly what you are not supposed
+to do... whats up?
```

Comment:
It's okay to return empty handles, just not handles that point to something without going through `EscapableHandleScope::Escape()`.